Make "--insecure-skip-tls-verify" work on all cases

rata opened the pull request:
In the getting started example of AWS, the master uses an IP that is changed on
stop/start. If you are playing with a cluster and stop and start the master,
the IP is changed and you can't connect again, even using the
"--insecure-skip-tls-verify" option.

This patch just removes the check to make it work in those cases.

I think there is no need to avoid it working, given the option name and
description and, if we want, even a bigger warning can be added. But not being
able to connect at all to a cluster with a master IP change seems like a really
bad user experience.

krousey commented:
@kubernetes/sig-api-machinery Any one see any issues with taking this check out?

ncdc commented:
cc @liggitt

liggitt commented on pkg/client/transport/transport.go:
```diff
@@ -55,9 +55,6 @@ func TLSConfigFor(c *Config) (*tls.Config, error) {
 	if !(c.HasCA() || c.HasCertAuth() || c.TLS.Insecure) {
 		return nil, nil
 	}
-	if c.HasCA() && c.TLS.Insecure {
```
we don't want to change it here, we'd want to make sure the CA was cleared when handling the `--insecure-skip-tls-verify` check

liggitt replied:
off the top of my head, in DirectClusterConfig.getCluster(), after merging all configs together:

```
	// an override of --insecure-skip-tls-verify=true and no accompanying CA/CA data should clear already-set CA/CA data
	// otherwise, a kubeconfig containing a CA reference would return an error that "CA and insecure-skip-tls-verify couldn't both be set"
	if config.overrides.ClusterInfo.InsecureSkipTLSVerify && len(config.overrides.ClusterInfo.CertificateAuthority) == 0 && len(config.overrides.ClusterInfo.CertificateAuthorityData) == 0 {
		mergedClusterInfo.CertificateAuthority = ""
		mergedClusterInfo.CertificateAuthorityData = nil
	}
```

rata replied:
Ohh, that seems way better! When I first started writing the patch I though something was setting it by error, but didn't found when it was written and I assumed maybe it was expected.  But what you say makes way more sense, will do that! :)

rata committed:
Make "--insecure-skip-tls-verify" work on all cases

In the getting started example of AWS, the master uses an IP that is changed on
stop/start. If you are playing with a cluster and stop and start the master,
the IP is changed and you can't connect again, even using the
"--insecure-skip-tls-verify" option.

This patch fixes it and makes the option work on those cases too by making
sure no CA/CAData is added when it shouldn't.

rata commented:
@liggitt does it look okay now? I've tested it and it seems to works fine.

Thanks again! :)

liggitt commented:
needs tests to make sure this works as expected

liggitt commented:
@deads2k you see any issues with this?

deads2k commented:
> @deads2k you see any issues with this?

No issues.  It does need a test though and we use camelcase, not `_`.

rata commented:
On Fri, Mar 18, 2016 at 04:53:35AM -0700, David Eads wrote:
> > @deads2k you see any issues with this?
> 
> No issues.  It does need a test though and we use camelcase, not `_`.

Great, I'll use camlcase and see where/how to add a test for this :-)

rata commented:
@deads2k: Sorry the delay. I think this should be ok now, but please let me know :)

deads2k commented:
@k8s-bot 
ok to test

deads2k commented on pkg/client/unversioned/clientcmd/client_config.go:
```diff
@@ -305,6 +305,14 @@ func (config *DirectClientConfig) getCluster() clientcmdapi.Cluster {
 	mergo.Merge(&mergedClusterInfo, config.overrides.ClusterInfo)
+	// An override of --insecure-skip-tls-verify=true and no accompanying CA/CA data should clear already-set CA/CA data
+	// otherwise, a kubeconfig containing a CA reference would return an error that "CA and insecure-skip-tls-verify couldn't both be set"
+	CALen := len(config.overrides.ClusterInfo.CertificateAuthority)
```
`caLen` to keep to conventions.

rata replied:
Right, thanks!

deads2k commented on pkg/client/unversioned/clientcmd/client_config_test.go:
```diff
@@ -96,6 +103,22 @@ func TestMergeContext(t *testing.T) {
 
 	matchStringArg(namespace, actual, t)
+
+	config = createCAValidTestConfig()
```
Please factor this into a separate test: `TestInsecureOverridesCA`

rata replied:
Ok, sure!

deads2k commented:
Minor comments, a squash, and then lgtm.

rata commented:
thanks a lot for the review!

rata commented:
@deads2k: fixed now. Let me know if something is missing to merge it

deads2k commented:
> @deads2k: fixed now. Let me know if something is missing to merge it

Squash to one commit please.  No reason I'd want the code without the test.

rata commented:
@deads2k: sorry, I understood to squash the changes you said. I squashed the 2 patches in one now :)

deads2k commented:
@rata you're tagged, but CI doesn't like formatting: `FAILED   hack/../hack/verify-gofmt.sh	6s`

rata commented:
@deads2k: grrr, sorry again :-S. I'm not on my regular notebook and it screwed it up. Fixed now :-/

rata commented:
The build says: "No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.". Should this be a travis problem?
